File: 백준/Gold/1956. 운동/운동.java
import java.io.*;
import java.util.*;

public class Main {
    public static final int INF = 1_000_000_000;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int v = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        int[][] dist = new int[v+1][v+1];

        for (int i = 1; i <= v; i++)
            for (int j = 1; j <= v; j++)
                dist[i][j] = INF;

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int sr = Integer.parseInt(st.nextToken()), en = Integer.parseInt(st.nextToken()),
                    c = Integer.parseInt(st.nextToken());
            dist[sr][en] = Math.min(dist[sr][en], c);
        }

        for (int k = 1; k <= v; k++)
            for (int i = 1; i <= v; i++)
                for (int j = 1; j <= v; j++)
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);

        int ans = INF * 2;
        for (int i = 1; i <= v; i++)
            for (int j = 1; j <= v; j++) {
                if (dist[i][j] == INF || dist[j][i] == INF) continue;
                if (i == j) ans = Math.min(ans, dist[i][j]);
                else ans = Math.min(ans, dist[i][j] + dist[j][i]);
            }

        System.out.println((ans == INF * 2) ? -1 : ans);
    }
}

File: 백준/Gold/1956. 운동/운동.cc
#include <bits/stdc++.h>
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

using namespace std;

const int INF = 1e9;

int dist[404][404], V, E;

int main(void) {
    fastio;
    cin >> V >> E;
    fill(&dist[0][0], &dist[0][0] + 404 * 404, INF);

    for (int i = 1; i <= E; i++) {
        int s, e, w;
        cin >> s >> e >> w;
        dist[s][e] = min(dist[s][e], w);
    }

    for (int k = 1; k <= V; k++)
        for (int i = 1; i <= V; i++)
            for (int j = 1; j <= V; j++)
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);


    int ans = INF * 2;
    for (int i = 1; i <= V; i++) {
        for (int j = 1; j <= V; j++) {
            if (dist[i][j] == INF || dist[j][i] == INF) continue;
            if (i == j) ans = min(ans, dist[i][j]);
            else ans = min(ans, dist[i][j] + dist[j][i]);
        }
    }

    if (ans == INF * 2) cout << -1;
    else cout << ans;

    return 0;
}